crates/wysiwyg/src/composer_model/selection.rs

```diff
@@ -62,6 +63,34 @@ where
             (s, e)
         }
     }
+
+    /// Return a boolean to let us know if we have a selection
+    pub fn has_selection(&self) -> bool {
+        let (s, e) = self.safe_selection();
+        s != e
+    }
+
+    /// Return a boolean to let us know if we have a cursor, ie a zero length selection
+    pub fn has_cursor(&self) -> bool {
+        let (s, e) = self.safe_selection();
+        s == e
+    }
+
+    /// Return a boolean to let us know if we are touching the edge of the DOM
+    pub fn has_selected_to_end_of_dom(&self) -> bool {
+        let (s, e) = self.safe_selection();
+        s == 0 || e == self.state.dom.text_len()
+    }
+
+    /// Return a boolean to let us know if we are touching the edge of the DOM
+    /// when moving in a specific direction
+    pub fn has_selected_end_of_dom(&self, direction: &Direction) -> bool {
+        let (s, e) = self.safe_selection();
+        match direction {
+            Direction::Forwards => e == self.state.dom.text_len(),
```

Comment:
I wish there was some way to 'cache' the text_len of Dom, as it could be a bit expensive to calculate it several times for complex trees, but I don't think we can do it easily. Just keep in mind trying to not use it often.

Reply:
Oh I didn't know that, that's good to know